grpclog/loggerv2_test.go

```diff
@@ -61,3 +66,231 @@ func checkLogForSeverity(s int, b []byte) error {
 	}
 	return nil
 }
+
+func TestLoggerV2PrintFunc(t *testing.T) {
+	// logFuncErr is a logger printFunc which generates a test error,
+	// for testing that the log func is not called
+	logFuncErr := func() string {
+		t.Errorf("want Logger func to not be called for io.Discard")
+		return ""
+	}
+
+	logFuncStr := "called-func"
+
+	// logFuncOk is a logger printFunc to write a constant string
+	logFuncOk := func() string {
+		return logFuncStr
+	}
+
+	// checkLogContainsFuncStr checks that the logger buffer logBuf contains the string written by logFuncOk
+	checkLogContainsFuncStr := func(logBuf []byte) {
+		if !bytes.Contains(logBuf, []byte(logFuncStr)) {
+			t.Errorf("want logger func to be called and print '%v', got '%v'", logFuncStr, string(logBuf))
+		}
+	}
+
+	// checkWrite checks that the log buffer buf was written as expected.
+	// The discard argument indicates there should not have been a write to the buffer.
+	// The logType indicates the type of the log, only used for better error messages.
+	// The msg is the string expected to have been written. Note we check "contains," since the log will have a prefix.
+	// The isJSON arg inddicates whether the logger was configured to write JSON.
+	checkWrite := func(buf *bytes.Buffer, discard bool, logType string, msg string, isJSON bool) {
+		bts, err := buf.ReadBytes('\n')
+		if discard {
+			if err == nil {
+				t.Fatalf("want discard %v to not write, got '%v'", logType, string(bts))
+			} else if err != io.EOF {
+				t.Fatalf("want discard %v buffer to be EOF, got '%v'", logType, err)
+			}
+		} else {
+			if err != nil {
+				t.Fatalf("want non-discard %v to not error, got '%v'", logType, err)
+			} else if !bytes.Contains(bts, []byte(msg)) {
+				t.Fatalf("want non-discard %v buffer contain message '%v', got '%v'", logType, msg, string(bts))
+			}
+			if isJSON {
+				obj := map[string]string{}
+				if err := json.Unmarshal(bts, &obj); err != nil {
+					t.Fatalf("want non-discard json %v to unmarshal, got '%v'", logType, err)
+				} else if _, ok := obj["severity"]; !ok {
+					t.Fatalf("want non-discard json %v to have severity field, got '%v'", logType, "missing")
+
+				} else if jsonMsg, ok := obj["message"]; !ok {
+					t.Fatalf("want non-discard json %v to have message field, got '%v'", logType, "missing")
+
+				} else if !strings.Contains(jsonMsg, msg) {
+					t.Fatalf("want non-discard json %v buffer contain message '%v', got '%v'", logType, msg, string(bts))
+				}
+			}
+		}
+	}
+
+	t.Run("discard only info", func(t *testing.T) {
```

Comment:
why are these tests isolated outside of for loop. We can move them to new tests with explicit names specifying the test case

Reply:
Changed